File: go/internal/handlers/health.go
package handlers

import (
	"context"
	"encoding/json"
	"net/http"

	"high_performance_api_benchmark/internal/database"
)

// Health returns liveness check.
func Health(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

// HealthTest returns custom health check (Bolt-compatible).
func HealthTest(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{
		"status":  "ok",
		"message": "Test health check endpoint",
	})
}

// Ready returns readiness check (DB). Returns 503 if unhealthy.
func Ready(w http.ResponseWriter, r *http.Request) {
	pool := database.Pool()
	if pool == nil {
		writeJSON(w, http.StatusServiceUnavailable, map[string]any{
			"status": "unhealthy",
			"checks": map[string]string{"database": "error"},
		})
		return
	}
	if err := pool.Ping(context.Background()); err != nil {
		writeJSON(w, http.StatusServiceUnavailable, map[string]any{
			"status": "unhealthy",
			"checks": map[string]string{"database": "error"},
		})
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"status": "healthy",
		"checks": map[string]string{"database": "ok"},
	})
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(v)
}

File: express/src/routes/health.js
/**
 * Health routes: /health, /health/test, /ready (Bolt-compatible).
 */

import { Router } from "express";
import { healthCheck } from "../db.js";

const router = Router();

router.get("/health", (req, res) => {
  res.json({ status: "ok" });
});

router.get("/health/test", (req, res) => {
  res.json({
    status: "ok",
    message: "Test health check endpoint",
  });
});

router.get("/ready", async (req, res) => {
  try {
    await healthCheck();
    res.json({ status: "healthy", checks: { database: "ok" } });
  } catch {
    res.status(503).json({
      status: "unhealthy",
      checks: { database: "error" },
    });
  }
});

export default router;
